### PR TITLE
Add new C APIs used by the new json release

### PR DESCRIPTION
**Pull request by chrisseaton**
Without these fixes essentially the entire Ruby ecosystem seems to stop working!

https://github.com/Shopify/truffleruby/issues/1

### CODE REVIEW

File: src/main/c/cext/ruby.c [hunk 1]
```diff
@@ -4715,7 +4715,7 @@ void rb_define_virtual_variable(const char *name, VALUE (*getter)(ANYARGS), void
 }
 
 void rb_gc_register_mark_object(VALUE obj) {
-  rb_tr_error("rb_gc_register_mark_object not implemented");
+  RUBY_CEXT_INVOKE("rb_gc_register_mark_object", obj);
```
Comment (eregon) [inline src/main/c/cext/ruby.c]: Minor: this could be `RUBY_INVOKE_NO_WRAP()` to avoid wrapping the return value.

**Comment by eregon** [comment]
Thanks!
@aardvark179 Could you review too?

> Without these fixes essentially the entire Ruby ecosystem seems to stop working!

I wonder what depends on the `json` gem and doesn't use the bundled `json`, maybe we should figure this out if it doesn't work as expected.

We could also ask for a new `json` release that checks if those C functions are available before using them, that way 19.3.0 would still work with latest `json`.

**Comment by Mogztter** [comment]
> We could also ask for a new json release that checks if those C functions are available before using them, that way TruffleRuby 19.3.0 would still work with latest json.

That would be great :+1:

**Comment by eregon** [comment]
> We could also ask for a new `json` release that checks if those C functions are available before using them, that way TruffleRuby 19.3.0 would still work with latest `json`.

That's unfortunately not so easy because we actually declare and define the function.
I'll check if we can avoid defining functions we don't implement, but I think we can't last time I tried.
Maybe we need something like `#ifdef GCCOMPACT` or so.

File: src/main/c/cext/ruby.c (hunk 1 above)
Comment (chrisseaton) [reply src/main/c/cext/ruby.c]: Done.

**Comment by chrisseaton** [comment]
> I wonder what depends on the `json` gem and doesn't use the bundled `json`

`i18n` is a concrete example at the bottom of many stacks. Non-standard-library `json` is everywhere.

File: src/main/c/cext/ruby.c (hunk 1 above)
Comment (chrisseaton) [reply src/main/c/cext/ruby.c]: You mean `RUBY_CEXT_INVOKE_NO_WRAP` I think.

Comment (eregon) [reply src/main/c/cext/ruby.c]: Yes, indeed.

**Pull request merged**